### ISSUE TITLE
shape mismatch of the forward function of the Loss class

### ISSUE THREAD

**Issue by Afreshbird** [zhulf0804/PointPillars#59]
According to your notes, all of the parameters's shape at zero index should be consistent in Loss class's forward function. But according to my dubug results, the shape of passed parameters is not the same.

According to your notes and my understanding, I made the following two changes：

1. In model/anchors.py     def anchor_target(batched_anchors, batched_gt_bboxes, batched_gt_labels, assigners, nclasses):
```
179    assigned_gt_labels_weights[pos_flag] = 1
180    assigned_gt_labels_weights[neg_flag] = 1
```
Be modified as:
`          assigned_gt_labels_weights[pos_flag] = 1`
After the revision, the shape at zero index of the bbox_cls_pred and batched_bbox_labels will be match with others variable. ( bbox_cls_pred and batched_bbox_labels at 109/180 and 111/182 row in train.py)

2.  In train.py 
```
108    num_cls_pos = (batched_bbox_labels < args.nclasses).sum()
109    bbox_cls_pred = bbox_cls_pred[batched_label_weights > 0]
110    batched_bbox_labels[batched_bbox_labels < 0] = args.nclasses
111    batched_bbox_labels = batched_bbox_labels[batched_label_weights > 0]
```
Be modified as:
```
          bbox_cls_pred = bbox_cls_pred[batched_label_weights > 0]
          batched_bbox_labels[batched_bbox_labels < 0] = args.nclasses
          batched_bbox_labels = batched_bbox_labels[batched_label_weights > 0]
          num_cls_pos = (batched_bbox_labels < args.nclasses).sum()
```
Just switched positions. Because batched_bbox_labels maybe contains the value of -1,  If num_cls_pos is assigned first, It can also lead to mismatched shapes.

And according to the issue of [#38](https://github.com/zhulf0804/PointPillars/issues/38), I made the following change in train.py.
```
102     # sin(a - b) = sin(a)*cos(b) - cos(a)*sin(b)
103     bbox_pred[:, -1] = torch.sin(bbox_pred[:, -1].clone()) * torch.cos(batched_bbox_reg[:, -1].clone())
104     batched_bbox_reg[:, -1] = torch.cos(bbox_pred[:, -1].clone()) * torch.sin(batched_bbox_reg[:, -1].clone())
```
Be modified as:
```
# sin(a - b) = sin(a)*cos(b) - cos(a)*sin(b)
a = bbox_pred[:, -1].clone()
b = batched_bbox_reg[:, -1].clone()
bbox_pred[:, -1] = torch.sin(a) * torch.cos(b)
batched_bbox_reg[:, -1] = torch.cos(a) * torch.sin(b)
```
The same goes for line 173-175.

**Comment by ZhangChenH** [comment]
Hello, 

1.  Thank you for your revision on the issue of #38,which has resolved my confusion.

2. Regarding the first question, my understanding is that in the `Loss` of the author's code, the `bbox cls loss` takes into account both `positive` samples and `negative` samples (background), while the `regression loss` and `direction cls loss` only use `positive` samples for calculation. Therefore, the shapes at the zero index positions of the former and the latter are different.

3. Regarding the second question,  `batched_bbox_labels` should not contain the value `-1`. This is because in line 176-177 of `model.anchors.py`,the values in `assigned_gt_labels` can only be within the range of `[0, nclasses]`.
```python
assigned_gt_labels = torch.zeros_like(cur_anchors[:, 0], dtype=torch.long) + nclasses
assigned_gt_labels[pos_flag] = gt_labels[assigned_gt_inds[pos_flag] - 1].long()
``` 

If my understanding is incorrect, please feel free to point it out.Thank you！
